### ISSUE TITLE
No supported way to get js doc tags from extraneous comments

### ISSUE THREAD

**Issue by rictic** [microsoft/TypeScript#42895]
### 🔎 Search Terms

ts.getJSDocTags
JSDoc tag not found

### 🕗 Version & Regression Information

Seen in TS 4.0 and 4.1

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about the compiler API

### 💻 Code

Consider the input file:

```ts
/** @foo */

/** @bar */

var x;
```

As far as I can tell, there's no supported compiler API to access the `@foo` jsdoc tag. I've resorted to casting ts.Node to `{jsDoc?: ts.JSDoc[]}`.

### 🙁 Actual behavior

ts.getJSDocTags and friends only give the `@bar` JSDoc tag

### 🙂 Expected behavior

That method, or another, to return both the `@foo` and `@bar` tags.

<details>
  <summary>Full compiler code sample to repro</summary>
  
```ts
import * as ts from 'typescript';

const filename = 'some_code.ts';
const program = ts.createProgram(['some_code.ts'], {noEmit: true});
program.emit();
const sourceFile = program.getSourceFile(filename);
if (!sourceFile) {
  throw new Error('could not get source file');
}

console.log(`getJSDocTags:`);
console.log([...getAllTags(sourceFile, (node) => ts.getJSDocTags(node))].map(
    (t) => t.tagName.text));
console.log(`Accessing jsDoc directly:`);
console.log(
    [...getAllTags(sourceFile, accessJsDoc)].map((t) => t.tagName.text));

function accessJsDoc(node: ts.Node): ts.JSDocTag[] {
  const tags = [];
  for (const jsDoc of (node as {jsDoc?: ts.JSDoc[]}).jsDoc ?? []) {
    for (const tag of jsDoc.tags ?? []) {
      tags.push(tag);
    }
  }
  return tags;
}

type TagGetter = (node: ts.Node) => Iterable<ts.JSDocTag>;
function getAllTags(
    node: ts.Node, getter: TagGetter, tags = new Set<ts.JSDocTag>()) {
  for (const tag of getter(node)) {
    tags.add(tag);
  }
  ts.forEachChild(node, (child) => {
    getAllTags(child, getter, tags);
  });
  return tags;
}

```

</details>

**Comment by RyanCavanaugh** [comment]
@sandersn is there (or should there be) an API for this?

**Comment by sandersn** [comment]
So far, we've tried to treat the `jsDoc` property as internal information so that people won't incorrectly miss the association between, for example, `x`'s `VariableDeclaration` and `@bar`, which is actually on the containing `VariableStatement`.

Generally there are two kinds of jsdoc: attached annotations (eg the `@bar` attached to `x`) and unattached declarations (eg `@typedef`).
Here's how the compiler works with the two kinds:

1. Look up jsdoc on nodes that are known to have attached-annotation jsdoc, using getJSDocTags or a more specific variant. Used for specific semantics.
2. Look up jsdoc on nodes that *might* have attached jsdoc, using `hasJSDocNodes(node: Node): node is HasJSDoc`, which is internal now. Used for generic purposes like tree walking or tree cleanup.
3. Walk the entire tree and treat unattached-declaration jsdoc nodes just like any other nodes (having obtained them as needed using `hasJSDocNodes`).

There's no safe, high-level API for getting unattached-declaration jsdoc tags. It's definitely a hole in our API that we don't -- but we haven't needed anything like it internally because we walk the AST all the time instead.

@rictic can you elaborate on the problem you're trying to solve? Are you trying to find all the declaration-like tags?

**Comment by rictic** [comment]
Consider:

```typescript
/** 
 * @requirecss {./styles}
 * @license
 * legal text legal text legal text
 */

/**
 * Comment describing Foo.
 */
export class Foo {}
```

Arguably this code is wrong, and one shouldn't have two successive jsdoc comments.

However, some jsdoc annotations like `@license` and `@fileoverview` are understood as documenting the file itself, not any node in that file.

Additionally, if multiple successive jsdoc comments are incorrect, I'd like to use the TypeScript API to emit a diagnostic in that case. For example, suppose that `@foo` and `@bar` are meaningful jsdoc tags that a tool is acting upon. I'd like to emit a diagnostic in this case, telling the user to collapse these two comments into one:

```typescript
/** @foo */
/** @bar */
var x;
```

**Comment by sandersn** [comment]
No, unattached-declaration jsdoc is legal and meaningful, we just don't have a high-level API to get it because we've never needed it.

File-initial jsdoc is so common that I think the best high-level API would be something that is, basically

```ts
function getFileInitialJSDoc(file: SourceFile) {
  if (file.statements.length && hasJSDocNodes(file.statements[0]) && file.statements[0].jsDoc.length > 1)
    return file.statements[0].jsDoc[0]
  // plus maybe a case for EndOfFileToken in case there aren't any statements in the file,
  // plus maybe return all but the last jsdoc from the first statement 
}
```

Then we could worry about a complex "find all unattached jsdoc" later, if it's ever needed.

**Comment by thorn0** [comment]
As for typedefs, it's annoying that whenever they're followed by a function declaration, TS attaches that docblock to this function and shows typedefs as part of the function's description whenever its name is hovered.

![image](https://user-images.githubusercontent.com/94334/113485775-085ae500-94b8-11eb-89d8-97cf5f4df1b9.png)

Same thing with variable declarations. Would be good to keep typedef-only docblocks unattached. Is this a known issue?

**Comment by sandersn** [comment]
So far I think we have two separate bugs:

1. Some API users want to find the file-initial jsdoc that commonly contains a header. Typescript doesn't understand file-level tags, so it's never needed an easy way to get this jsdoc. [My comment above](https://github.com/microsoft/TypeScript/issues/42895#issuecomment-785490401) gives a partial design for that function with a couple of unanswered questions. @rictic does the above function look like it meets your needs?
2. Declaration-like jsdocs are incorrectly attached to the following statement. The fix is probably as simple as a filter somewhere, but I'm not sure which component should do the filtering. @thorn0 can you file a new bug for the above so we can track and discuss it separately?

**Comment by thorn0** [comment]
> can you file a new bug

Done: https://github.com/microsoft/TypeScript/issues/43534
